### PR TITLE
feat: add prober readme

### PR DESCRIPTION
**Pull request by sailorlqh**
fixes: #177

### CODE REVIEW

File: docs/prober-and-monitoring.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
+# Prober and monitoring
+
+**PMAP is not a official Google product.**
+
+Monitoring consists of two parts: prober and alerting.
+
+## Prober
+
+[Prober](../prober/) is a go binary being deployed to [cloud run
```
Comment (capri-xiyue) [inline docs/prober-and-monitoring.md:9]: Explain what prober does first and then talk about the technical details like cloud run job and cloud scheduler.

File: docs/prober-and-monitoring.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+job](https://cloud.google.com/run/docs/overview/what-is-cloud-run#jobs), and is
+triggered by [cloud scheduler](https://cloud.google.com/scheduler) to constantly
+probing pmap services to check if the services are up.
+
+In each prober job, the prober will upload a yaml file to
```
Comment (capri-xiyue) [inline docs/prober-and-monitoring.md:14]: Prober doesn't upload ONE yaml file. You can explain it based on CUJ. One CUJ is policy and the other CUJ is resource mapping

File: docs/prober-and-monitoring.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+on if the following requirements for both mapping and policy service are met:
+
+1. The object is uploaded.
+2. There is an entry in bigquery table that matches the traceID.
+
+## Monitoring and Alert Policies
+
+We monitor the pmap service with [native cloud run monitoring
+metrics](https://cloud.google.com/monitoring/api/metrics_gcp#gcp-run) and set up
+the alert for all pubsub subscriptions base on the following metrics:
```
Comment (capri-xiyue) [inline docs/prober-and-monitoring.md:26]: add pubsub metrics link

File: docs/prober-and-monitoring.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+a corrosponding entry exists in bigquery table. The job is considered as success
+on if the following requirements for both mapping and policy service are met:
+
+1. The object is uploaded.
+2. There is an entry in bigquery table that matches the traceID.
+
+## Monitoring and Alert Policies
+
+We monitor the pmap service with [native cloud run monitoring
+metrics](https://cloud.google.com/monitoring/api/metrics_gcp#gcp-run) and set up
```
Comment (capri-xiyue) [inline docs/prober-and-monitoring.md:25]: Which metrics do we use here for cloud run, do you mean the `job/completed_execution_count`?

File: docs/prober-and-monitoring.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+You can use the provided Terraform module to setup Prober, or you can refer to
+the provided module to build it from scratch.
+
+```terraform
+module "prober_and_monitoring" {
+  source = "../modules/monitoring"
+
+  project_id = var.project_id
+
+  prober_bucket_id           = "GCS bucket id for where objects are uploaded to."
```
Comment (capri-xiyue) [inline docs/prober-and-monitoring.md:46]: Make it clear which variables are users that need to provide. Currently, it doesn't look clear.

**Review by capri-xiyue — CHANGES_REQUESTED** [review]

File: docs/prober-and-monitoring.md (hunk 5 above)
Comment (sailorlqh) [reply docs/prober-and-monitoring.md:57]: Updated

File: docs/prober-and-monitoring.md (hunk 3 above)
Comment (sailorlqh) [reply docs/prober-and-monitoring.md]: added

File: docs/prober-and-monitoring.md (hunk 2 above)
Comment (sailorlqh) [reply docs/prober-and-monitoring.md]: updated.

File: docs/prober-and-monitoring.md (hunk 1 above)
Comment (sailorlqh) [reply docs/prober-and-monitoring.md]: Added

File: docs/prober-and-monitoring.md [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+[Prober](../prober/) is a tool that constantly probes pmap services to check if
+the services are up. It is built as a go binary and deployed to [cloud run
+job](https://cloud.google.com/run/docs/overview/what-is-cloud-run#jobs), and is
+triggered by [cloud scheduler](https://cloud.google.com/scheduler) to constantly
+probing pmap services to check if the services are up.
+
+In each prober job, the prober will cover two CUJ:
+
+- Resource mapping (mapping service user journey)
+- Wipeout planning (policy service user journey)
```
Comment (yolocs) [inline docs/prober-and-monitoring.md:18]:
```suggestion
In each prober execution, the prober will cover two CUJ:

- Import resource mapping to BigQuery from GCS
- Import policies to BigQuery from GCS
```

File: docs/prober-and-monitoring.md [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+monitor pubsub subsrciption and sent out alerts. The metrics used are:
+
+- subscription/oldest_unacked_message_age
+- subscription/dead_letter_message_counts
+
+We also use [cloud run
+metircs](https://cloud.google.com/monitoring/api/metrics_gcp#gcp-run) to monitor
+and send alert base prober job execution result. The metric used is:
+
+- job/completed_execution_count
```
Comment (yolocs) [inline docs/prober-and-monitoring.md:41]: Mention this before pubsub. And then mention why we additionally need to watch the pubsub metrics

File: docs/prober-and-monitoring.md [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+- Resource mapping (mapping service user journey)
+- Wipeout planning (policy service user journey)
+
+For each CUJ, prober will upload a yaml file to
+[gcs](https://cloud.google.com/storage), and query the bigquery table to see if
+a corrosponding entry exists in bigquery table. The job is considered as success
+on if the following requirements for both mapping and policy service are met:
+
+1. The object is uploaded.
+2. There is an entry in bigquery table that matches the traceID.
```
Comment (yolocs) [inline docs/prober-and-monitoring.md:26]:
```suggestion
1. The object is uploaded.
2. There is an entry in bigquery table that matches the traceID.

### Limitation

Ideally the prober should cover the full user journey which includes checking in
resource mappings and policies in a GitHub repository. However, there is no
effective way to automate PR authoring and bypass branch protection. As a result,
we seek the next closest journey - upload probing files to GCS bucket directly.
```

**Review by yolocs — COMMENTED** [review]
Some wording suggestion

File: docs/prober-and-monitoring.md (hunk 7 above)
Comment (sailorlqh) [reply docs/prober-and-monitoring.md:41]: Added

**Review by yolocs — DISMISSED** [review]

File: docs/prober-and-monitoring.md [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+**PMAP is not a official Google product.**
+
+Monitoring consists of two parts: prober and alerting.
+
+## Prober
+
+[Prober](../prober/) is a tool that constantly probes pmap services to check if
+the services are up. It is built as a go binary and deployed to [cloud run
+job](https://cloud.google.com/run/docs/overview/what-is-cloud-run#jobs), and is
```
Comment (capri-xiyue) [inline docs/prober-and-monitoring.md:11]: Nits: `It is built as a go binary, deployed as [cloud run
job](https://cloud.google.com/run/docs/overview/what-is-cloud-run#jobs), and 
triggered by [cloud scheduler](https://cloud.google.com/scheduler).`

File: docs/prober-and-monitoring.md [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+triggered by [cloud scheduler](https://cloud.google.com/scheduler) to constantly
+probing pmap services to check if the services are up.
+
+In each prober execution, the prober will cover two CUJ:
```
Comment (capri-xiyue) [inline docs/prober-and-monitoring.md:15]: nits: CUJs

File: docs/prober-and-monitoring.md [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+prober separately. Otherwise, you can use the following terraform code to set up
+your prober. All these variables are required.
+
+```terraform
+module "prober_and_monitoring" {
+  source = "../modules/monitoring"
+
+  project_id = var.project_id
+
+  prober_bucket_id           = "GCS bucket id for where objects are uploaded to."
```
Comment (capri-xiyue) [inline docs/prober-and-monitoring.md:65]: prober_bucket_id           = "<YOUR_GCS_bucket_id_for_where_objects_are_uploaded to>" Just wanna make it clear that what variables that users need to provide.

File: docs/prober-and-monitoring.md (hunk 9 above)
Comment (sailorlqh) [reply docs/prober-and-monitoring.md]: Added

File: docs/prober-and-monitoring.md (hunk 10 above)
Comment (sailorlqh) [reply docs/prober-and-monitoring.md]: fixed

File: docs/prober-and-monitoring.md (hunk 11 above)
Comment (sailorlqh) [reply docs/prober-and-monitoring.md]: done

**Review by capri-xiyue — APPROVED** [review]

**Review by yolocs — APPROVED** [review]

**Pull request merged**